src/components/views/elements/MemberEventListSummary.js

```diff
@@ -144,87 +144,105 @@ module.exports = React.createClass({
         );
     },
 
-    render: function() {
-        let summary = null;
+    shouldComponentUpdate: function(nextProps, nextState) {
+        return (
+            nextProps.events.length !== this.props.events.length ||
+            nextState.expanded !== this.state.expanded
+        );
+    },
 
-        // Reorder events so that joins come before leaves
+    render: function() {
         let eventsToRender = this.props.events;
+        let fewEvents = eventsToRender.length < this.props.threshold;
+        let expanded = this.state.expanded || fewEvents;
+
+        let expandedEvents = null;
+        if (expanded) {
+            expandedEvents = this.props.children;
+        }
+
+        if (fewEvents) {
+            return (
+                <div className="mx_MemberEventListSummary">
+                    {expandedEvents}
+                </div>
+            );
+        }
+
+        // Map user IDs to the first and last member events in eventsToRender for each user
+        let userEvents = {
+            // $userId : {first : e0, last : e1}
+        };
+
+        eventsToRender.forEach((e) => {
+            const userId = e.getSender();
+            // Initialise a user's events
+            if (!userEvents[userId]) {
+                userEvents[userId] = {first: null, last: null};
+            }
+            if (!userEvents[userId].first) {
+                userEvents[userId].first = e;
+            } else {
+                userEvents[userId].last = e;
```

Comment:
why not make this unconditional and remove the `if(!lastEvent) { lastEvent = firstEvent; }` at line 202?